### PR TITLE
[Perf][Kernel] Fused DSA indexer Top-k kernel (LiteTopk)

### PR DESCRIPTION
**Pull request by Heisenberg-Yin**
## 📌 Description

Hi @youkaichao @WoosukKwon @mgoin @hmellor @njhill @Isotr0py @DarkLight1337, could you please take a look at this PR when you have a chance?

This PR introduces two complementary SM100 prefill optimizations for DeepSeek Sparse Attention (DSA):

- **LiteTopK** accelerates the indexer by fusing MQA scoring, online filtering, and TopK selection without materializing the `[num_q, seq_len]` score matrix.
- **LiteDSA** further accelerates the subsequent sparse-attention computation by packing multiple neighboring queries into one larger masked attention GEMM, reusing their heavily overlapping KV sets.

To the best of our knowledge, LiteTopk is the first Indexer-Topk implementation that avoids to store the score matrix, and LiteDSA is the first that converts multiple small per-token sparse-attention GEMMs into a grouped masked GEMM while preserving each query's original attend set. The corresponding paper is available on [arXiv](https://arxiv.org/abs/2607.11976).

**At 1M-token prefill, LiteTopK accelerates GLM-5.2 and DeepSeek-V4-Flash by 1.29× and 1.08×, respectively. LiteDSA increases the end-to-end speedups to 1.40× and 1.25×, with no accuracy loss.**

Thanks to my collaborators Jianyang Gao ([@gaoj0017](https://github.com/gaoj0017)) from ETH Zurich, Peiqi Yin ([@yinpeiqi](https://github.com/yinpeiqi)) from CUHK, and Jiangneng Li ([@gravesprite](https://github.com/gravesprite)) from NTU.

### Motivation

DSA models such as **DeepSeek-V3.2/DeepSeek-V4**, **GLM-4.6/GLM-5.2**, and **LongCat-2.0** use a sparse-attention indexer during prefill. The current indexer computes an FP8 MQA score for every query/KV pair, materializes a `[num_q, seq_len]` matrix in HBM, and then runs a per-row TopK:

```python
logits = fp8_fp4_mqa_logits(...)        # materializes [num_q, seq_len] in HBM
ops.top_k_per_row_prefill(logits, ...)  # TopK over the materialized matrix
```

At 256K–1M context, this transient matrix is both large and expensive to write and read. After the indexer, the sparse-attention path launches many narrow per-token attention GEMMs. Neighboring queries usually select substantially overlapping KV positions, so those kernels repeatedly load and process much of the same KV data; at high tensor-parallel degrees, padding a small number of local heads to the kernel tile shape also wastes tensor-core work.

### LiteTopK: fused indexer scoring and TopK

LiteTopK avoids storing the score matrix:

1. Score KV tiles with FP8 MQA using tcgen05 UMMA.
2. Maintain an online **bucketed gate** with 256 histogram buckets and a self-calibrating threshold, discarding unlikely positions as scores are produced.
3. Run a compact TopK over the surviving candidates.

The candidate workspace is bounded independently of `seq_len`, replacing the dense `O(num_q × seq_len)` logits allocation and its extra HBM pass.

### LiteDSA: grouped masked sparse attention

LiteDSA targets the attention stage after TopK. If a rank owns `H` query heads, it groups `G = 128 / H` adjacent query tokens and packs their `G × H` real query/head rows into one 128-row tensor-core tile. For example, GLM-5.2 TP8 packs 16 neighboring tokens per attention call. Instead of running `G` small sparse GEMMs, LiteDSA performs one larger GEMM over the union of their selected KV positions.

The grouped computation remains query-exact:

1. **Union and KV reuse.** A GPU bitmap kernel deduplicates the group's TopK lists, emits the union in ascending order, and fuses logical-to-physical block-table conversion. Neighboring queries exhibit strong causal locality; the GLM path observes roughly 6–7× KV deduplication.
2. **Per-query masking.** A query-major membership bitmask records which union entries belong to each original query. The masked FP8 attention kernel applies this mask before softmax, so every query attends to exactly its original TopK set.
3. **Lower orchestration overhead.** Union, membership, and output buffers are persistent overwrite-write buffers. The union plan is version-cached and reused across layers that share the same indexer output, avoiding repeated allocation, clearing, and reconstruction.

DeepSeek-V4's C128A path uses the same packing principle with a structure-aware specialization. At TP8, 16 tokens × 8 real heads fill one 128-row tile instead of padding each token from 8 to 64 heads. Its compressed-prefix plus sliding-window structure is represented by exact per-query ranges over the group union; the ranges are derived directly from positions, cached per chunk, and avoid materializing the `[num_tokens, topk + window]` combined-index matrix. The measured union size is only 1.002× that of a single token.

In both paths, each union KV entry is loaded once for the group rather than repeatedly for every neighboring query. Unsupported shapes, CUDA Graph capture, non-SM100 devices, and other ineligible configurations transparently fall back to the existing attention path.

### Configuration

```bash
# LiteTopK indexer + existing sparse attention
VLLM_LITETOPK=1 VLLM_DSA_MODE=litetopk vllm serve <dsa-model> ...

# LiteTopK indexer + LiteDSA grouped sparse attention (supported FP8 sparse MLA path)
VLLM_LITETOPK=1 VLLM_DSA_MODE=litedsa vllm serve <dsa-model> ...

# DeepSeek-V4 C128A head-packed attention specialization
VLLM_DSV4_PACKED_ATTN=1 vllm serve <deepseek-v4-model> ...
```

## 🧪 Test Plan

> The kernel and end-to-end checks require an SM100 (B200) build and a DSA model.

```bash
# Build for SM100
uv pip install -e . --torch-backend=auto

# LiteDSA orchestration, buffer-lifetime, and DeepSeek-V4 packed-path tests
.venv/bin/python -m pytest -q \
    tests/model_executor/layers/test_litedsa.py \
    tests/model_executor/layers/test_dsv4_packed_attn.py

# End-to-end long-context parity and performance
VLLM_LITETOPK=1 VLLM_DSA_MODE=litetopk \
    vllm serve <dsa-model> --max-model-len ...
VLLM_LITETOPK=1 VLLM_DSA_MODE=litedsa \
    vllm serve <dsa-model> --max-model-len ...

# Model evaluation for output parity
lm_eval --model vllm --model_args pretrained=<dsa-model>,...  # compare raw/litetopk/litedsa
```

### LiteTopK kernel-level validation

Hardware: NVIDIA B200 (SM100). Inputs: real GLM-5.2 indexer tensors, `Q=8192`, `top_k=2048`.

The dense FP32 logits buffer consumes **8 GiB at 256K, 16 GiB at 512K, 24 GiB at 768K, and 32 GiB at 1M**. LiteTopK removes this allocation.

| KV length | Dense (ms) | LiteTopK (ms) | Paired speedup median | Recall median |
|---:|---:|---:|---:|---:|
| 262,144 | 13.0469 | 10.6117 | 1.2266× | 99.9977% |
| 524,288 | 25.8997 | 20.4716 | 1.2652× | 99.9980% |
| 786,432 | 39.2471 | 30.0235 | 1.3084× | 99.9978% |
| 1,048,576 | 51.8106 | 40.1796 | 1.2929× | 99.9978% |

Recall rounds to **100.00%** at two decimal places across the tested settings.

### LiteDSA kernel-level validation

On captured DeepSeek-V4 C128A production data, head packing reduces sparse-attention kernel time from **6.565 ms to 0.839 ms (7.83×)**. Validation reports LSE max absolute error of `3e-6` and output relative error of `0.0025`.

### End-to-end 1M-token prefill

| Model | Input tokens | Dense (s) | LiteTopK (s) | LiteTopK speedup | LiteDSA (s) | LiteDSA speedup |
|---|---:|---:|---:|---:|---:|---:|
| GLM-5.2 TP8+EP8 | 1M | 152.89 | 118.66 | 1.29× | 109.14 | 1.40× |
| DeepSeek-V4 TP4+EP | 1M | 60.00 | 55.7 | 1.08× | 48.1 | 1.25× |

## Author

Ziqi Yin, Nanyang Technological University

### CODE REVIEW

**Comment by MatthewBonanni** [comment]
cc @LopezCastroRoberto

**Review by LopezCastroRoberto — COMMENTED** [review]
Thanks for the contribution. Review in progress. One important thing missing is a model-level accuracy eval; the PR currently provides no results supporting the “no accuracy loss” claim.

**Review by LopezCastroRoberto — COMMENTED** [review]
I recommend running some similar MRCR 2-needle eval as https://github.com/vllm-project/vllm/pull/43008 for different context-lens vs upstream

**Comment by Heisenberg-Yin** [comment]
> I recommend running some similar MRCR 2-needle eval as #43008 for different context-lens vs upstream

# ContextRaw @ 2GB Benchmark

| Context Length | Raw | LiteTopK | LiteDSA |
|---:|---:|---:|---:|
| 128K | 8.0958s / 1.00x | 8.0185s / **1.010x** | 6.9346s / **1.167x** |
| 256K | 18.7273s / 1.00x | 17.8909s / **1.047x** | 15.7599s / **1.188x** |
| 512K | 48.1018s / 1.00x | 43.3118s / **1.111x** | 39.0873s / **1.231x** |
| 1M | 142.7266s / 1.00x | 115.8325s / **1.232x** | 107.2425s / **1.331x** |

**Comment by LopezCastroRoberto** [comment]
@Heisenberg-Yin what are these numbers exactly? Time needed to execute the eval? I think this is secondary, we need the accuracy scores to make sure everything is correct. Thanks!

**Comment by Heisenberg-Yin** [comment]
> similar MRCR 2-needle eval as #43008

The above time is the prefill time of different sequence.

The MRCR 2-needle eval scores is as below:

# MRCR 2-needle results

| 长度 | n | Raw | LiteTopK | LiteDSA |
| --- | ---: | ---: | ---: | ---: |
| 32K–64K | 50 | 0.883458 | 0.902042 | 0.941533 |
| 64K–128K | 50 | 0.889256 | 0.870255 | 0.941472 |
| 128K–256K | 20 | 0.815006 | 0.859008 | 0.902167 |
| 256K–512K | 10 | 0.807099 | 0.899956 | 0.806545 |
| 512K–1M | 5 | 0.631580 | 0.621226 | 0.631580 |

**Review by LopezCastroRoberto — COMMENTED** [review]
Thank you for the accuracy numbers @Heisenberg-Yin! Could you clarify why LiteTopK and LiteDSA differ by up to around 9 percentage points in some buckets? I can’t tell whether it’s statistically significant from these numbers alone, but it seems a bit large to dismiss as noise. Were they run on the same prompts with deterministic settings?

Also, would you consider splitting this into multiple PRs? LiteTopK, generic LiteDSA, and the DeepSeek-V4 packed-attention specialization have largely independent implementation and validation paths. The PR adds about 23K lines. Separate PRs, each including its tests and benchmarks, would make the code and performance/correctness claims substantially easier to review.

**Comment by Heisenberg-Yin** [comment]
> Thank you for the accuracy numbers @Heisenberg-Yin! Could you clarify why LiteTopK and LiteDSA differ by up to around 9 percentage points in some buckets? I can’t tell whether it’s statistically significant from these numbers alone, but it seems a bit large to dismiss as noise. Were they run on the same prompts with deterministic settings?
> 
> Also, would you consider splitting this into multiple PRs? LiteTopK, generic LiteDSA, and the DeepSeek-V4 packed-attention specialization have largely independent implementation and validation paths. The PR adds about 23K lines. Separate PRs, each including its tests and benchmarks, would make the code and performance/correctness claims substantially easier to review.

Dear [LopezCastroRoberto](https://github.com/LopezCastroRoberto)

I’ll integrate only LiteTopK into this repo, minimize the number of files changed, and present the reason of the 2-needle test results.

Best,
Ziqi

**Comment by Heisenberg-Yin** [comment]
> > Thank you for the accuracy numbers @Heisenberg-Yin! Could you clarify why LiteTopK and LiteDSA differ by up to around 9 percentage points in some buckets? I can’t tell whether it’s statistically significant from these numbers alone, but it seems a bit large to dismiss as noise. Were they run on the same prompts with deterministic settings?
> > Also, would you consider splitting this into multiple PRs? LiteTopK, generic LiteDSA, and the DeepSeek-V4 packed-attention specialization have largely independent implementation and validation paths. The PR adds about 23K lines. Separate PRs, each including its tests and benchmarks, would make the code and performance/correctness claims substantially easier to review.
> 
> Dear [LopezCastroRoberto](https://github.com/LopezCastroRoberto)
> 
> I’ll integrate only LiteTopK into this repo, minimize the number of files changed, and present the reason of the 2-needle test results.
> 
> Best, Ziqi

## Summary

Dear @LopezCastroRoberto

This PR updates the LiteTopK implementation used by vLLM.

With the updated implementation, enabling LiteTopK no longer affects the
2-needle retrieval score. Across all evaluated context-length ranges, LiteTopK
matches the native baseline within normal measurement variation, including at
contexts up to 1M tokens.

## 2-Needle Evaluation

| Model | Method | 64K--128K | 128K--256K | 256K--512K | 512K--1M |
|:------|:-------|-----------:|------------:|------------:|----------:|
| GLM-5.2 | Native | 96.10 | 90.86 | 0.902542 | 63.16 |
| GLM-5.2 | LiteTopK | 96.11 | 90.89 | 0.902498 | 63.12 |

## 1M-Token Performance

At a sequence length of 1,048,320 tokens (approximately 1M), LiteTopK reduces
the runtime from 86 seconds to 49 seconds, delivering a **1.73×
speedup**.

| Model | Context Length | Native (s) | TopK (s) | Speedup |
|---|---:|---:|---:|---:|
| GLM-5.2 | 256K | 12.87 | 11.06 | **1.16×** |
| GLM-5.2 | 512K | 31.08 | 22.96 | **1.35×** |
| GLM-5.2 | 768K | 55.70 | 35.90 | **1.55×** |
| GLM-5.2 | 1M | 86.03 | 49.72 | **1.73×** |
| DeepSeek-V4-Flash | 256K | 6.41 | 6.10 | **1.05×** |
| DeepSeek-V4-Flash | 512K | 17.06 | 12.74 | **1.34×** |
| DeepSeek-V4-Flash | 768K | 32.05 | 20.91 | **1.53×** |
| DeepSeek-V4-Flash | 1M | 51.22 | 30.63 | **1.67×** |
| Hy-4-preview | 256K | 13.76 | 12.02 | **1.14×** |
| Hy-4-preview | 512K | 32.78 | 24.93 | **1.32×** |
| Hy-4-preview | 768K | 57.45 | 38.86 | **1.48×** |
| Hy-4-preview | 1M | 87.92 | 53.81 | **1.63×** |

The maximum observed difference from the native baseline is 0.04 points. These
results show that the updated LiteTopK path preserves the model's 2-needle
retrieval quality and introduces no measurable accuracy regression.

We also cleans up parts of the implementation to keep the diff focused and reduce the number of changed lines and files.
